modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/GlobalSnapshotEventCutter.scala

```diff
@@ -0,0 +1,144 @@
+import org.tessellation.ext.kryo.RefinedSerializer
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.node.shared.domain.event.EventCutter
+import org.tessellation.schema.Block
+import org.tessellation.schema.gossip.Ordinal
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.security.hash._
+import org.tessellation.security.{Hashed, Hasher}
+
+import eu.timepit.refined.types.numeric.PosInt
+
+object GlobalSnapshotEventCutter {
+  type GlobalPeerIdEventMap = Map[PeerId, List[(Ordinal, GlobalSnapshotEvent)]]
+
+  def make[F[_]: Async: KryoSerializer: Hasher](maxBinarySizeBytes: PosInt): EventCutter[F, PeerId, Ordinal, GlobalSnapshotEvent] =
+    new EventCutter[F, PeerId, Ordinal, GlobalSnapshotEvent] {
+
+      def cut(events: GlobalPeerIdEventMap): F[GlobalPeerIdEventMap] =
+        needsCutting(events)
+          .ifM(
+            OrderedEvents(events)
+              .map(Iterator.unfold(_)(_.next))
+              .flatMap(removeEvents(events, _)),
+            events.pure[F]
+          )
+
+      def needsCutting(events: GlobalPeerIdEventMap): F[Boolean] =
+        events.toBinary.liftTo[F].map(_.length > maxBinarySizeBytes.value)
+
+      def removeEvents(startingEvents: GlobalPeerIdEventMap, iterator: Iterator[PeerIdEvent]): F[GlobalPeerIdEventMap] =
+        startingEvents.tailRecM { events =>
+          for {
+            eventToRemove <- MonadThrow[F].fromOption(iterator.nextOption(), new IllegalStateException("No GlobalSnapshotEvent events"))
+            reducedEvents = removeOneEvent(events, eventToRemove)
+            cutAgain <- needsCutting(reducedEvents)
+          } yield Either.cond(!cutAgain, reducedEvents, reducedEvents)
+        }
+
+      def removeOneEvent(events: GlobalPeerIdEventMap, eventToRemove: PeerIdEvent): GlobalPeerIdEventMap =
+        events.updatedWith(eventToRemove.id) {
+          case None => None
+          case Some(globalPeerIdEvent) =>
+            val globalSnapshotEvent = eventToRemove.toGlobalSnapshotEvent
+            val updatedEvents = globalPeerIdEvent.filterNot { case (_, gse) => gse == globalSnapshotEvent }
+            Option.when(updatedEvents.nonEmpty)(updatedEvents)
+        }
+
+      trait PeerIdEvent {
+        def id: PeerId
+        def fee: Long
+        def hash: Hash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent
+      }
+
+      val ordering: Ordering[PeerIdEvent] = Ordering.by[PeerIdEvent, (Long, Hash)](e => (e.fee, e.hash))
+
+      case class PeerIdStateChannelEvent(id: PeerId, event: StateChannelEvent) extends PeerIdEvent {
+        val fee: Long = event.snapshotBinary.value.fee.value.value
+        def hash: Hash = event.snapshotBinary.lastSnapshotHash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent = event.asLeft[DAGEvent]
+      }
+
+      case class PeerIdHashedEvent(id: PeerId, event: Hashed[Block]) extends PeerIdEvent {
```

Comment:
It's a `...HashedBlock` event or `...HashedDagBlock` event. I don't think `HashedEvent` is clear and also it's inconsistent with the naming of `PeerIdStateChannelEvent`

Reply:
You could even drop the `Hashed` I guess as the types require `Hashed` either way